src/main/java/org/broadinstitute/hellbender/tools/exome/sexgenotyper/PloidyAnnotatedTargetCollection.java

```diff
@@ -0,0 +1,219 @@
+package org.broadinstitute.hellbender.tools.exome.sexgenotyper;
+
+import com.google.common.collect.Sets;
+import htsjdk.samtools.util.Locatable;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.tools.exome.Target;
+import org.broadinstitute.hellbender.tools.exome.TargetCollection;
+import org.broadinstitute.hellbender.utils.IndexRange;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+
+import javax.annotation.Nonnull;
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * A collection of {@link Target} instances along with helper methods for generating their genotype ploidy
+ * annotations on-the-fly using provided contig ploidy annotations.
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+public final class PloidyAnnotatedTargetCollection implements TargetCollection<Target> {
+
+    /**
+     * Map from targets to their ploidy annotations (based on target contigs)
+     */
+    private Map<Target, ContigPloidyAnnotation> targetToContigPloidyAnnotationMap;
+
+    /**
+     * List of autosomal targets
+     */
+    private List<Target> autosomalTargetList;
+
+    /**
+     * List of allosomal targets
+     */
+    private List<Target> allosomalTargetList;
+
+    /**
+     * List of all targets
+     */
+    private List<Target> fullTargetList;
+
+    /**
+     * Set of all targets
+     */
+    private Set<Target> fullTargetSet;
```

Comment:
The `Set` version is used for checking whether a given target is present or not (it provides an `O(1)` contains(...) method). The `List` version is used for preserving indexing. There is a `LinkedHashSet<Target>` which is a set implementation that preserves the order but it does not provide `O(1)` index lookup, which we want. Another approach is to have the Set version and a `Map<Target,Integer>` for storing the indices, but it is slightly slower. Also, we are not dealing with large objects here (Target is a very small class), so it is OK to store the targets in multiple containers for different purposes.